File: 2024/10/10.rb
#!/usr/bin/env ruby

DELTAS = [[0, -1], [0, 1], [-1, 0], [1, 0]].freeze

def find_trailheads(grid)
  trailheads = []
  grid.each_with_index do |row, y|
    row.each_with_index do |height, x|
      trailheads << [x, y] if height == 0
    end
  end
  trailheads
end

def max_trails(grid, trailhead, unique_paths = false)
  trails = 0
  visited = Set.new
  paths = [trailhead]
  x_range, y_range = [0...grid[0].size, 0...grid.size]

  while paths.any?
    x, y = paths.shift
    next if !unique_paths and visited.include? [x, y]
    visited << [x, y]
    height = grid[y][x]

    if height == 9
      trails += 1
      next
    end

    DELTAS.each do |dx, dy|
      nx, ny = [x + dx, y + dy]
      paths << [nx, ny] if x_range.include?(nx) and y_range.include?(ny) and grid[ny][nx] == height + 1
    end
  end

  trails
end

grid = []
$stdin.readlines.each_with_index do |line|
  grid << line.chomp.chars.map(&:to_i)
end

total1, total2 = [0, 0]

find_trailheads(grid).each do |trailhead|
  total1 += max_trails(grid, trailhead, false)
  total2 += max_trails(grid, trailhead, true)
end

puts total1, total2

File: 2024/10/10.go
package main

import (
	"bufio"
	"os"
)

type P struct {
	x int
	y int
}

func (this *P) add(other *P) P {
	return P{this.x + other.x, this.y + other.y}
}

type Grid []Row
type Row []rune

var DELTAS []P = []P{{0, -1}, {0, 1}, {-1, 0}, {1, 0}}

func parseInput() *Grid {
	grid := Grid{}
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		grid = append(grid, Row(scanner.Text()))
	}
	return &grid
}

func findTrailheads(grid *Grid) []P {
	trailheads := []P{}
	for y, row := range *grid {
		for x, cell := range row {
			if cell == '0' {
				trailheads = append(trailheads, P{x, y})
			}
		}
	}
	return trailheads
}

func inBounds(p *P, xmax, ymax int) bool {
	return p.x >= 0 && p.y >= 0 && p.x <= xmax && p.y <= ymax
}

func maxTrails(grid *Grid, trailhead P, uniquePaths bool) int {
	trails := 0
	visited := map[P]bool{}
	paths := []P{trailhead}
	var p P
	xmax, ymax := len((*grid)[0])-1, len(*grid)-1

	for len(paths) > 0 {
		p, paths = paths[0], paths[1:]

		if !uniquePaths && visited[p] {
			continue
		}
		visited[p] = true
		height := (*grid)[p.y][p.x]

		if height == '9' {
			trails++
			continue
		}

		for _, d := range DELTAS {
			np := p.add(&d)
			if inBounds(&np, xmax, ymax) && (*grid)[np.y][np.x] == height+1 {
				paths = append(paths, np)
			}
		}
	}

	return trails
}

func main() {
	grid := parseInput()
	total1, total2 := 0, 0
	for _, trailhead := range findTrailheads(grid) {
		total1 += maxTrails(grid, trailhead, false)
		total2 += maxTrails(grid, trailhead, true)
	}
	println(total1)
	println(total2)
}
